dftimewolf/lib/collectors/gsheets.py

```diff
@@ -0,0 +1,243 @@
+            name='{0:s}_{1:s}'.format(spreadsheet_title, sheet_title),
+            path=output_path)
+        self.state.StoreContainer(sheet_csv_file)
+
+    except (RefreshError, DefaultCredentialsError) as exception:
+      self.ModuleError('Something is wrong with your gcloud access token or '
+                       'Application Default Credentials. Try running:\n '
+                       '$ gcloud auth application-default login')
+      self.ModuleError(exception, critical=True)
+
+  def _GetCredentials(self) -> Credentials:
+    """Obtains API credentials for accessing the Google Sheets API.
+
+    Returns:
+      google.oauth2.credentials.Credentials: Google API credentials.
+    """
+    credentials = None
+
+    # The credentials file stores the user's access and refresh tokens, and is
+    # created automatically when the authorization flow completes for the first
+    # time.
+    credentials_path = os.path.join(
+        os.path.expanduser('~'), self._CREDENTIALS_FILENAME)
+    lock = filelock.FileLock(credentials_path + '.lock')  # pylint: disable=abstract-class-instantiated
+    with lock:
+      if os.path.exists(credentials_path):
+        try:
+          credentials = Credentials.from_authorized_user_file(
+              credentials_path, self.SCOPES)
+        except ValueError as exception:
+          self.logger.warning(
+              'Unable to load credentials: {0:s}'.format(exception))
+          credentials = None
+
+      # If there are no (valid) credentials available, let the user log in.
+      if not credentials or not credentials.valid:
+        if credentials and credentials.expired and credentials.refresh_token:
+          credentials.refresh(Request())
+        else:
+          secrets_path = os.path.join(
+              os.path.expanduser('~'), self._CLIENT_SECRET_FILENAME)
+          if not os.path.exists(secrets_path):
+            error_message = (
+                'No OAuth application credentials available to access google '
+                'sheets. Please generate OAuth application credentials (see '
+                'https://developers.google.com/sheets/api/guides/authorizing) '
+                'and save them to {0:s}.').format(secrets_path)
+            self.ModuleError(error_message, True)
+          flow = InstalledAppFlow.from_client_secrets_file(
+              secrets_path, self.SCOPES)
+          credentials = flow.run_console()
+
+        # Save the credentials for the next run
+        with open(credentials_path, 'w') as token_file:
+          token_file.write(credentials.to_json())
+
+    return credentials
+
+  def _BuildSheetsResource(self, credentials: Credentials) -> discovery.Resource:
+    """Builds a Google Sheets resource object to use to request logs.
+
+    Args:
+      credentials: Google API credentials
+
+    Returns:
+      A resouce object for interacting with the Google Sheets API.
+    """
+    resource = discovery.build('sheets', 'v4', credentials=credentials)
+
```

Comment:
Please remove the newline here, and maybe return the resource directly?

Reply:
Donw